clippy_lints/src/manual_clamp.rs

```diff
@@ -0,0 +1,720 @@
+        input.zip(min.zip(max)).map(|(input, (min, max))| ClampSuggestion {
+            params: InputMinMax { input, min, max },
+            span: expr.span,
+            add_semicolon: false,
+        })
+    }
+    // Lots of expensive queries coming up which we don't want to repeat for every expr, so short
+    // circuit this a bit.
+    if let ExprKind::Call(
+        Expr { kind: ExprKind::Path(_), .. },
+        &[ref arg_first, _]
+    ) = &simplify_kind(expr).kind
+    && is_clampable(cx, cx.typeck_results().expr_ty(arg_first)) {
+        if let ExprKind::Call(
+            Expr { kind: ExprKind::Path(_), .. },
+            args_inner
+        ) = &simplify_kind(arg_first).kind
+        && args_inner.len() == 2 {
+            // Matched, do nothing
+        }
+        else if let ExprKind::Call(
+            Expr { kind: ExprKind::Path(_), .. },
+            &[_, ref arg_second]
+        ) = &simplify_kind(expr).kind
+        && let ExprKind::Call(
+            Expr { kind: ExprKind::Path(_), .. },
+            args_inner
+        ) = &simplify_kind(arg_second).kind
+        && args_inner.len() == 2 {
+            // Matching, do nothing
+        } else {
+            return None;
+        }
+    } else {
+        return None;
+    }
+    let core_max_res = def_path_res(cx, &["core", "cmp", "max"]);
+    let core_ord_max_res = def_path_res(cx, &["core", "cmp", "Ord", "max"]);
+    let core_min_res = def_path_res(cx, &["core", "cmp", "min"]);
+    let core_ord_min_res = def_path_res(cx, &["core", "cmp", "Ord", "min"]);
+    let f32_min_res = def_path_res(cx, &["f32", "min"]);
+    let f32_max_res = def_path_res(cx, &["f32", "max"]);
+    let f64_min_res = def_path_res(cx, &["f64", "min"]);
+    let f64_max_res = def_path_res(cx, &["f64", "max"]);
+    for_methods(cx, expr, false, core_min_res, core_max_res)
+        .or_else(|| for_methods(cx, expr, true, core_max_res, core_min_res))
+        .or_else(|| for_methods(cx, expr, true, core_ord_max_res, core_ord_min_res))
+        .or_else(|| for_methods(cx, expr, false, core_ord_min_res, core_ord_max_res))
+        .or_else(|| for_methods(cx, expr, true, f32_max_res, f32_min_res))
+        .or_else(|| for_methods(cx, expr, false, f32_min_res, f32_max_res))
+        .or_else(|| for_methods(cx, expr, true, f64_max_res, f64_min_res))
+        .or_else(|| for_methods(cx, expr, false, f64_min_res, f64_max_res))
+    // In theory you could use the core::cmp methods with the Ord methods, though this seems pretty
+    // unlikely
+}
+
+/// Targets patterns like
+///
+/// ```
+/// # let (input, min, max) = (0, -3, 12);
+///
+/// match input {
+///     input if input > max => max,
+///     input if input < min => min,
+///     input => input,
+/// }
+/// # ;
+/// ```
+fn is_match_pattern<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) -> Option<ClampSuggestion<'tcx>> {
+    if let ExprKind::Match(value, arms, rustc_hir::MatchSource::Normal) = &simplify_kind(expr).kind
+    && arms.len() == 3
```

Comment:
Iterating enables us to not care about the order of the arms here, why would we want to remove it?

Reply:
They always will be in the same order unless I'm misunderstanding what you mean, as in

```rust
match input {
    n => n,
    // these two arms are dead code, since the first arm always matches
    n if n > max => max,
    n if n < min => min,
}
```

Reply:
Oh, no I just forgot that the thru arm has to be last. Good point, I'll implement it.

Reply:
Fixed in [`4fe8643`](https://github.com/rust-lang/rust-clippy/pull/9484/commits/4fe8643e02c0163bac6c21659f5ebb150af6817b)